### PR TITLE
[gtasks] Add basic support for Google Tasks

### LINKED ISSUE #21: Sync as Google Tasks
It would be great to have options to be synced as Google Tasks, or Google Reminder, rather than as events

### PR DESCRIPTION
**Pull request by kisseliov**
# Description

Added basic an integration with Google Tasks API to enable syncing Taskwarrior tasks as Google Tasks rather than events.

Fixes #21

## Type of change

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

# How Has This Been Tested?

Added tests similar to other integrations: conversions between Google Task and Taskwarrior task formats, creating, completing, deleting, annotating tasks.

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by kstreser** [comment]
I like this very much! You are my hero. Thank you.

**Comment by bergercookie** [comment]
This works looks great and apologies I have not reviewed it sooner.

I'll be trying it out this weekend and I'll hopefully provide some feedback.

Thanks!

File: syncall/google/gtasks_side.py [hunk 1]
```diff
@@ -0,0 +1,294 @@
...
+            assert isinstance(dt, datetime.datetime)
+            return dt
+
+    @staticmethod
+    def format_datetime(dt: datetime.datetime) -> str:
+        assert isinstance(dt, datetime.datetime)
+        return format_datetime_tz(dt)
+
+    @classmethod
+    def parse_datetime(cls, dt: Union[str, dict, datetime.datetime]) -> datetime.datetime:
```
Comment (bergercookie) [inline syncall/google/gtasks_side.py:227]: This looks like a duplicate of the function we have in GCalSide. Could we abstract this? Perhaps put it in the parent class?

File: syncall/tw_gtasks_utils.py [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+        date_key = "due"
+
+    end_time = GTasksSide.get_task_completed_time(gtasks_item)
+    tw_item[date_key] = end_time
+
+    # update time
+    if "updated" in gtasks_item.keys():
+        tw_item["modified"] = GTasksSide.parse_datetime(gtasks_item["updated"])
+
+    # NOTE(kisseliov): consider adding custom UDAs for Google Tasks to handle positions and hierarchy
```
Comment (bergercookie) [inline syncall/tw_gtasks_utils.py:127]: Are you still planning on implementing this?

File: syncall/tw_gtasks_utils.py [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+    return tw_item
+
+
+def _parse_gtask_notes(
```
Comment (bergercookie) [inline syncall/tw_gtasks_utils.py:131]: Is this similar to the one we currently have for parsing google calendar tasks? Perhaps abstract this if it is?

File: syncall/types.py [hunk 4]
```diff
@@ -40,6 +40,73 @@ class TaskwarriorRawItem(TypedDict, total=False):
...
+            * uuid: 542f5dbc-b1b7-4a85-b55e-10f5f1d31847
+        ''',
+        'due': '2021-12-04T18:07:00.000Z',
+        "completed": "2021-12-04T15:07:00.000Z",
+        'links': []
+    }"""
+
+    completed: Optional[
+        str
+    ]  # Completion date of the task (as a RFC 3339 timestamp). This field is omitted if the task has not been completed.
```
Comment (bergercookie) [inline syncall/types.py:89]: nitpick: Could you put the comment at the previous line from the variable? I find it to be a bit easier to read that way. i.e.:


    # Completion date of the task (as a RFC 3339 timestamp). This field is omitted if the task has not been completed.
    completed: Optional[str]

File: tests/conftest_gtasks.py [hunk 5]
```diff
@@ -0,0 +1,157 @@
+from typing import cast
```
Comment (bergercookie) [inline tests/conftest_gtasks.py:1]: Thanks for adding tests as well! 💯 💯 💯

File: pyproject.toml [hunk 6]
```diff
@@ -74,6 +76,7 @@ rfc3339 = "^6.2"
 typing = "^3.7.4"
 item-synchronizer = "^1.1.4"
 bubop = { version = "0.1.10", allow-prereleases = true }
+google-auth-httplib2 = "^0.1.0"
```
Comment (bergercookie) [inline pyproject.toml:79]: Is this required for all the integrations? If it is google-dependent could you mark it as optional and add it to the `google` extra [here](https://github.com/bergercookie/syncall/pull/79/files#diff-50c86b7ed8ac2cf95bd48334961bf0530cdc77b5a56f852c5c61b89d735fd711R82)?

File: readme-tw-gtasks.md [hunk 7]
```diff
@@ -0,0 +1,87 @@
+# [Taskwarrior](https://taskwarrior.org/) ⬄ [Google Tasks](https://support.google.com/tasks/answer/7675772?hl=en&co=GENIE.Platform%3DDesktop)
```
Comment (bergercookie) [inline readme-tw-gtasks.md:1]: Could you remove the part after `?`, i.e., keep just the `https://support.google.com/tasks/answer/7675772`

File: readme-tw-gtasks.md [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+
+## Description
+
+Given all tasks in your Google Task task list of a
```
Comment (bergercookie) [inline readme-tw-gtasks.md:5]: minor corrections: Given all tasks in your Google Tasks *and the* task list of a taskwarrior _filter_ ...

**Review by bergercookie — COMMENTED** [review]
Hi @kisseliov and sorry about the very late review.

All in all this is some great functionality! 👍🏾 🥳 . I haven't delved into it exhaustively but from a first look here's a few things to adress before merging:

* I noticed that a completion of a task in google tasks result in a removal in taskwarrior. Is this intentional? This creates problems when tasks are marked as completed in Gtasks and then moved back to pending (again in gtasks): `AssertionError: Item already has a UUID, try updating it instead of adding it`. Could we instead have it so that a completion of a google task leads to a completion of the respective taskwarrior task?
* Let's pt a higher-res gif - use something like gifski to generate it. I've used it with a command like this in the past: `gifski --fast-forward 1.5 -o test.gif video.mp4`
* When creating the app in the Developer Console, the user needs to specifically add a bunch of scopes for the Google Tasks API to access:  `Create, edit, organize, and delete all your tasks`  and `View your tasks`. Could this be part of the readme notes?
* Could we add a note in the tw<>gcal integration to redirect users to this integration instead? I think for the majority of cases a google task is what users would want - instead of a calendar event.


I'm also posting below a list of checks I usually do when I'm adding a new integration (sub XXX, YYY for Taskwarrior and Google Tasks :) ). if you have some time to go through them I 'd really appreciate, otherwise, no worries, I can do it on my own time.

* [x] Verify you can add items from XXX and they appear in YYY during the initial run with both sides being empty
* [x] Verify you can add items from XXX and they appear in YYY during the initial run with YYY being empty at first
* [x] Verify you can add items from YYY and they appear in XXX during the initial run with XXX being empty at first
* [x] Verify you can update the description of items from XXX and the changes appear in YYY
* [ ] Verify you can update the description of items from YYY and the changes appear in XXX
* [x] Verify you can mark items as completed from XXX and they get completed in YYY
* [x] Verify you can mark items as completed from YYY and they get completed in XXX
* [ ] Verify you can delete items from XXX and they get deleted in YYY
* [ ] Verify you can delete items from YYY and they get deleted in XXX
* [ ] Verify you can create *completed* items from XXX and they appear completed in YYY
* [ ] Verify you can create *completed* items from YYY and they appear completed in XXX
* [ ] Verify you can create *deleted* items from XXX and they appear deleted in YYY
* [ ] Verify you can create *deleted* items from YYY and they appear deleted in XXX
* [ ] Verify you can create in-progress items from XXX and they appear as in-progress/started in YYY
* [x] Verify you can create in-progress items from YYY and they appear as in-progress/started in XXX

Finally, putting this here so as to not forget, let's go through the following checks before merging:

* [x] Rebase branch on top of latest master
* [x] Resolve any conflicts
* [x] Squash all commits into one

**Comment by bergercookie** [comment]
Could you also add a line in [generate-shell-completions.sh](https://github.com/kisseliov/syncall/blob/tw-gtask-sync/scripts/generate-shell-completions.sh) and call this script so as to generate new completions?

**Comment by kisseliov** [comment]
@bergercookie All good! I'll work on it when I have some free time. :)

**Comment by bergercookie** [comment]
Hi @kisseliov ,

Did you have time to work on this? If not I can look at resolving some of the pending issues tomorrow.

File: pyproject.toml (hunk 6 above)
Comment (bergercookie) [reply pyproject.toml]: done

File: syncall/types.py (hunk 4 above)
Comment (bergercookie) [reply syncall/types.py]: done

**Comment by bergercookie** [comment]
> Could you also add a line in [generate-shell-completions.sh](https://github.com/kisseliov/syncall/blob/tw-gtask-sync/scripts/generate-shell-completions.sh) and call this script so as to generate new completions?

That's done now

**Comment by kisseliov** [comment]
@bergercookie Hi! Not really, I would appreciate if you help :)

File: syncall/tw_gtasks_utils.py (hunk 2 above)
Comment (bergercookie) [reply syncall/tw_gtasks_utils.py]: Keeping the todo around for now but not doing anything about it.

File: syncall/tw_gtasks_utils.py (hunk 3 above)
Comment (bergercookie) [reply syncall/tw_gtasks_utils.py]: Created a new function and called it from both the gtasks and gcal paths.

File: readme-tw-gtasks.md (hunk 7 above)
Comment (bergercookie) [reply readme-tw-gtasks.md]: done in latest revision

**Comment by bergercookie** [comment]
> I noticed that a completion of a task in google tasks result in a removal in taskwarrior. Is this intentional? This creates problems when tasks are marked as completed in Gtasks and then moved back to pending (again in gtasks): AssertionError: Item already has a UUID, try updating it instead of adding it. Could we instead have it so that a completion of a google task leads to a completion of the respective taskwarrior task?

> When creating the app in the Developer Console, the user needs to specifically add a bunch of scopes for the Google Tasks API to access: Create, edit, organize, and delete all your tasks and View your tasks. Could this be part of the readme notes?

> Could we add a note in the tw<>gcal integration to redirect users to this integration instead? I think for the majority of cases a google task is what users would want - instead of a calendar event.

All three points are now complete.

**Comment by bergercookie** [comment]
> Let's pt a higher-res gif - use something like gifski to generate it. I've used it with a command like this in the past: gifski --fast-forward 1.5 -o test.gif video.mp4

Also completed

**Comment by bergercookie** [comment]
Thank you so much for this work @kisseliov. 💯 💯 💯 
I'll be merging this to main and soon I'll push a new version to pypi as well.

**Pull request merged**